keps/sig-auth/3299-kms-v2-improvements/README.md

```diff
@@ -242,34 +243,79 @@ message StatusResponse {
 }
 ```
 
-The `key_id` may be funneled into the storage version status as another field that API servers can attempt to gain consensus on:
+### Rotation
 
-```diff
-diff --git a/staging/src/k8s.io/api/apiserverinternal/v1alpha1/types.go b/staging/src/k8s.io/api/apiserverinternal/v1alpha1/types.go
-index bfa249e135c..e671fe599a9 100644
---- a/staging/src/k8s.io/api/apiserverinternal/v1alpha1/types.go
-+++ b/staging/src/k8s.io/api/apiserverinternal/v1alpha1/types.go
-@@ -56,6 +56,8 @@ type StorageVersionStatus struct {
-	 // +optional
-	 CommonEncodingVersion *string `json:"commonEncodingVersion,omitempty" protobuf:"bytes,2,opt,name=commonEncodingVersion"`
- 
-+    CommonKeyID *string `json:"commonKeyID,omitempty" protobuf:"bytes,4,opt,name=commonKeyID"`
-+
-	 // The latest available observations of the storageVersion's state.
-	 // +optional
-	 // +listType=map
-@@ -77,6 +79,8 @@ type ServerStorageVersion struct {
-	 // The encodingVersion must be included in the decodableVersions.
-	 // +listType=set
-	 DecodableVersions []string `json:"decodableVersions,omitempty" protobuf:"bytes,3,opt,name=decodableVersions"`
-+
-+    KeyID *string `json:"keyID,omitempty" protobuf:"bytes,4,opt,name=keyID"`
- }
- 
- type StorageVersionConditionType string
+The `key_id` will be funneled into the status of the new encryption state REST API.  Each API server will create this resource (via a controller that will likely run as a post start hook) with a random name on startup (stale objects will be garbage collected by a new controller that runs in the Kubernetes controller manager).  This resource provides the necessary information for a controller to automatically initiate storage migration when consensus has been reached among the different API server instances.  Opaque hashes are used to prevent leaking any specific information about the encryption configuration and the key IDs in use.  This is a cluster scoped REST API that is only meant to be read by the associated controllers and the cluster admin.
+
+**Question: should writes to this resource be limited to the API server loopback identity?**
+
+**Question: should we have a status field with the Kubernetes version to prevent migration during differences in API server version?**
+
+```go
+// metadata.name is the ID of the API server
```

Comment:
Can we use this ID (https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/1965-kube-apiserver-identity)?

Reply:
Sadly no because it is alpha and has not made any progress in ~2 years.  If it ever does go GA, we could start using it since the values just need to be unique and are otherwise opaque.

Reply:
could we add this as a note [here](https://github.com/kubernetes/enhancements/tree/master/keps/sig-auth/3299-kms-v2-improvements#alternatives)?